exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/SubsetTransformer.java

```diff
@@ -45,15 +48,20 @@ public RelTraitSet newTraitSet(RelTrait... traits) {
 
   }
 
-  boolean go(T n, RelNode candidateSet) throws E {
+  public boolean go(T n, RelNode candidateSet) throws E {
     if ( !(candidateSet instanceof RelSubset) ) {
       return false;
     }
 
     boolean transform = false;
+    Set<RelNode> transformedRels = Sets.newHashSet();
     for (RelNode rel : ((RelSubset)candidateSet).getRelList()) {
       if (isPhysical(rel)) {
         RelNode newRel = RelOptRule.convert(candidateSet, rel.getTraitSet().plus(Prel.DRILL_PHYSICAL));
+        if(transformedRels.contains(newRel)) {
```

Comment:
Change to use newIdentityHashSet. 

And we could not simply mark a node is isTransformed, since there could be many rules transforming the same node. Maintaining a map of applied rules in relNode might not be a good idea since remembering what rule has been applied should not be the job of a relNode(set). So what about maintaining such map in planner? Not a good idea since the same node might need to be converted into different sets. So I would think leaving this part as is.